src/hal/src/pso/mod.rs

```diff
@@ -221,32 +248,112 @@ pub struct SpecializationConstant {
 }
 
 /// Specialization information structure.
-#[derive(Debug, Copy)]
-pub struct Specialization<'a> {
+#[derive(Debug, Copy, Clone)]
+pub struct RawSpecialization<'a> {
     /// Constant array.
     pub constants: &'a [SpecializationConstant],
     /// Raw data.
     pub data: &'a [u8],
 }
 
-impl<'a> Default for Specialization<'a> {
+impl RawSpecialization<'_> {
+    /// Create empty specialization instance.
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl Default for RawSpecialization<'_> {
     fn default() -> Self {
-        Specialization {
+        RawSpecialization {
             constants: &[],
             data: &[],
         }
     }
 }
 
-impl<'a> Clone for Specialization<'a> {
-    fn clone(&self) -> Self {
-        Specialization {
-            constants: self.constants,
-            data: self.data,
+/// Convert typed specialization into untyped one.
+pub trait Specialization {
+    /// Build `Specialization` value.
+    fn as_raw(&self) -> RawSpecialization<'_>;
+}
+
+/// Temporary storage for specialization constants.
+#[derive(Clone, Debug, Default)]
+pub struct SpecializationStorage {
+    /// Actual data for specialization constants.
+    pub data: Vec<u8>,
+
+    /// Specialization constants that referer to the `data`.
+    pub constants: Vec<SpecializationConstant>,
+}
+
+impl SpecializationStorage {
+    /// Create empty specialization.
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl Specialization for SpecializationStorage {
+    fn as_raw(&self) -> RawSpecialization<'_> {
+        RawSpecialization {
+            data: &self.data,
+            constants: &self.constants,
         }
     }
 }
 
+/// Empty `Specialization` implementation.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Unspecialized;
```

Comment:
can this just be something like `RawSpecialization::EMPTY`?

Reply:
I forgot associated constants exists. How could I?